File: app/src/main/java/com/lazysoul/kotlinwithandroid/singletons/TodoManager.java
package com.lazysoul.kotlinwithandroid.singletons;

import com.lazysoul.kotlinwithandroid.datas.Todo;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by Lazysoul on 2017. 7. 12..
 */

public class TodoManager {

    public static final String KEY_ID = "id";

    public static final String KEY_BODY = "body";

    public static final String KEY_REQUEST_TYPE = "request_type";

    public static final String KEY_RESULT_TYPE = "result_type";

    public static final int REQUEST_TYPE_CREATE = 100;

    public static final int REQUEST_TYPE_VIEW = 101;

    public static final int RESULT_TYPE_CREATED = 200;

    public static final int RESULT_TYPE_UPDATED = 201;

    private static ArrayList<Todo> todoList = new ArrayList<>();

    public static List<Todo> getTodoList() {
        return todoList;
    }

    public static ArrayList<Todo> createSamples() {
        for (int i = 0; i < 10; i++) {
            Todo todo = new Todo();
            todo.setId(i);
            todo.setChecked(false);
            todo.setBody("Todo " + i);
            todoList.add(todo);
        }
        return todoList;
    }

    public static int getMaxId() {
        int max = -1;
        for (Todo todo : todoList) {
            if (todo.getId() > max) {
                max = todo.getId();
            }
        }
        return max;
    }

    @Nullable
    public static Todo getTodo(int id) {
        for (Todo todo : todoList) {
            if (todo.getId() == id) {
                return todo;
            }
        }
        return null;
    }

    @NonNull
    public static ArrayList<Todo> search(String text) {
        ArrayList<Todo> result = new ArrayList<>();
        if (text.isEmpty()) {
            result.addAll(todoList);
        } else {

            for (Todo todo : todoList) {
                if (todo.getBody().contains(text)) {
                    result.add(todo);
                }
            }
        }
        return result;
    }

    @NonNull
    public static Todo insert(int id, String body) {
        Todo todo = new Todo();
        todo.setId(id);
        todo.setBody(body);
        todo.setChecked(false);
        todoList.add(todo);
        return todo;
    }

    @Nullable
    public static Todo update(int id, String body) {
        for (Todo todo : todoList) {
            if (todo.getId() == id) {
                todo.setBody(body);
                return todo;
            }
        }
        return null;
    }

    public static void checked(int id, boolean isChecked) {
        for (Todo todo : todoList) {
            if (todo.getId() == id) {
                todo.setChecked(isChecked);
                break;
            }
        }
    }
}

File: app/src/main/kotlin/com/lazysoul/kotlinwithandroid/singletons/TodoManager.kt
package com.lazysoul.kotlinwithandroid.singletons

import com.lazysoul.kotlinwithandroid.datas.Todo

import java.util.ArrayList

/**
 * Created by Lazysoul on 2017. 7. 12..
 */

object TodoManager {

    const val KEY_ID = "id"

    const val KEY_BODY = "body"

    const val KEY_REQUEST_TYPE = "request_type"

    const val KEY_RESULT_TYPE = "result_type"

    const val REQUEST_TYPE_CREATE = 100

    const val REQUEST_TYPE_VIEW = 101

    const val RESULT_TYPE_CREATED = 200

    const val RESULT_TYPE_UPDATED = 201

    private val todoList = ArrayList<Todo>()

    val maxId: Int
        get() = todoList.maxBy { it.id }?.id ?: -1

    fun getTodoList(): List<Todo> = todoList

    fun createSamples(): ArrayList<Todo> {
        (0..9).mapTo(todoList) {
            Todo(it).apply {
                isChecked = false
                body = "Todo " + it
            }
        }
        return todoList
    }

    fun getTodo(id: Int): Todo = todoList.firstOrNull { it.id == id } ?: throw  NullPointerException("todo is null")

    fun search(text: String): ArrayList<Todo> {
        val result = ArrayList<Todo>()
        if (text.isEmpty()) {
            result.addAll(todoList)
        } else {
            todoList.filterTo(result) { it.body.contains(text) }
        }
        return result
    }

    fun insert(id: Int, body: String): Todo =
        Todo(id, body, true)
            .also { todoList.add(it) }

    fun update(id: Int, body: String): Todo? = todoList
        .firstOrNull { it.id == id }
        ?.also { it.body = body }

    fun checked(id: Int, isChecked: Boolean) {
        todoList.firstOrNull { it.id == id }
            ?.also { it.isChecked = isChecked }
    }
}
